shopify_integration/oauth.py

```diff
@@ -0,0 +1,133 @@
+import binascii
+import json
+import os
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+from shopify.session import Session as ShopifySession
+from shopify.utils.shop_url import sanitize_shop_domain
+
+import frappe
+from frappe import _
+
+if TYPE_CHECKING:
+	from frappe.integrations.doctype.connected_app.connected_app import ConnectedApp
+	from frappe.integrations.doctype.token_cache.token_cache import TokenCache
+	from shopify_integration.shopify_integration.doctype.shopify_settings.shopify_settings import (
+		ShopifySettings,
+	)
+
+
+@frappe.whitelist()
+def initiate_web_application_flow(settings: str):
+	"""Return an authorization URL for the user, and save the state in Token Cache."""
+
+	if isinstance(settings, str):
+		settings: Dict = json.loads(settings)
+
+	shopify_settings: "ShopifySettings" = frappe.get_doc(
+		"Shopify Settings", settings.get("name")
+	)
+	connected_app: "ConnectedApp" = frappe.get_doc(
+		"Connected App", settings.get("connected_app")
+	)
+
+	ShopifySession.setup(
+		api_key=connected_app.client_id,
+		secret=connected_app.get_password("client_secret"),
+	)
+
+	state = binascii.b2a_hex(os.urandom(15)).decode("utf-8")
+	shopify_session = ShopifySession(
+		shopify_settings.shopify_url, shopify_settings.api_version
+	)
+	auth_url = shopify_session.create_permission_url(
+		connected_app.get_scopes(), connected_app.redirect_uri, state
+	)
+
+	# create an initial token cache for the user
+	token_cache = connected_app.get_token_cache(frappe.session.user)
+	if not token_cache:
+		token_cache = frappe.new_doc("Token Cache")
+		token_cache.user = frappe.session.user
+		token_cache.connected_app = connected_app.name
+	token_cache.state = state
+	token_cache.save(ignore_permissions=True)
+
+	return auth_url
+
+
+@frappe.whitelist(allow_guest=True)
+def callback(*args, **kwargs):
+	validate_request(*args, **kwargs)
+	connected_app, token_cache = get_oauth_details(*args, **kwargs)
+	shopify_settings = get_shopify_settings(connected_app)
+
+	if not shopify_settings:
+		frappe.throw(_(f"No Shopify Settings found for {kwargs.get('shop')}"))
+
+	# remove Frappe's argument to properly validate HMAC
+	kwargs.pop("cmd")
+
+	session = ShopifySession(shopify_settings.shopify_url, shopify_settings.api_version)
+	access_token = session.request_token(kwargs)
+	token_cache = token_cache.update_data(
+		{
+			"token_type": "Bearer",
+			"access_token": access_token,
+		}
+	)
+
+	frappe.local.response["type"] = "redirect"
+	frappe.local.response["location"] = (
+		token_cache.get("success_uri") or connected_app.get_url()
+	)
+
+
+def validate_request(*args, **kwargs):
+	# validate request method
+	if frappe.request.method != "GET":
+		frappe.throw(_("Invalid request method: {}").format(frappe.request.method))
+
+	# validate redirect URI parameters
+	path = frappe.request.path[1:].split("/")
```

Comment:
what

Reply:
Don't ask lol. This is how they extract the connected app name from the request path.

Reply:
Just add a comment explaining this spaghetti

Reply:
Done